paddle/fluid/framework/operator.cc

```diff
@@ -1175,6 +1175,34 @@ proto::VarType::Type OperatorWithKernel::IndicateDataType(
   return data_type;
 }
 
+proto::VarType::Type OperatorWithKernel::IndicateVarDataType(
+    const ExecutionContext& ctx, const std::string& name) const {
+  const std::vector<const Variable*> vars = ctx.MultiInputVar(name);
+  PADDLE_ENFORCE_EQ(
+      vars.empty(), false,
+      "The %s Op doesn't have Input %s or it is an empty Variable.",
+      this->type_, name);
+  const Variable* var = vars.front();
+  const Tensor* t = nullptr;
+  if (var->IsType<LoDTensor>()) {
+    t = &var->Get<LoDTensor>();
+  } else if (var->IsType<SelectedRows>()) {
+    t = &(var->Get<SelectedRows>().value());
+  } else {
+    PADDLE_THROW(
+        "The Input Variable(%s) of %s Op used to determine kernel data type "
+        "should be LoDTensor or SelectedRows.",
+        name, this->type_);
+  }
+  // No need to check if t is nullptr, if so, var->isType or var->Get will throw
+  // error.
+  PADDLE_ENFORCE_EQ(
+      t->IsInitialized(), true,
+      "The Tensor in the %s Op's Input Variable %s is not initialized.",
+      this->type_, name);
+  return t->type();
+}
+
```

Comment:
Some logic of `OperatorWithKernel::IndicateDataType(const ExecutionContext& ctx, const std::string& name)` and `OperatorWithKernel::IndicateDataType(const ExecutionContext& ctx)` are the same, maybe this part can be shared.

Reply:
agree, I try to extract public parts of two functions.